### ISSUE TITLE
Switchable Hotkey Profiles

### ISSUE THREAD

**Issue by DingusShingleton** [endless-sky/endless-sky#2291]
So I have played many a game that uses WASD over the years, and I thought it might be interesting to try to rebind the entire set of ES hotkeys into such a format. I basically set it up so that the left hand controls all movement and navigation with WASD, and the right hand controls weapons and targeting using the arrow keys. I found this (after the initial adjustment period of course) to be a pretty cool and streamlined way to play (with the added bonus of removing the readjustment period when I come back from overwatch ><).

My suggestion is this: many games offer the ability to cycle through different keybinding arrangements (south paw, boxer, recon, from the halo days). I think this would be a cool idea in ES and allow people to try different combinations that might suit them better. For me, I prefer to arrange keys based on ease of use rather than matching the letter to the function (IE "F" is fight, "G" is gather, "M" is map, "J" is jump). Yes, they are easier to remember, but ultimately this system is slower and less efficient imo. A profile swap function would cater to everyone's preference.

**Comment by MessyMix** [comment]
IMO if this is going to be implemented it's going to have to be community made. If you can source, then I would suggest that you do it yourself, because although I think it's an interesting addition, it's going to be very low on the priority list (just think about it, MZ has more important things to do than to increase ease of life by a small notch). If someone does make it well, though, it would be a good change, provided that MZ approves.

I would help, but I can't C++ that well :3

**Comment by DingusShingleton** [comment]
I'm in the same boat. Honestly my hopes aren't super high (I didn't expect MZ to do it in the first place (unless it just so happened to pique his interest)), but I figured I would at least get some feedback/see if there was interest.

**Comment by KiLEdEnNis** [comment]
at least make it that A & D don't skip in-space messages (because whenever I get 1 while I'm steering I auto skip it)
I WASD'd my setting at the start & after keybindings became a thing had to switch secondary weapons to X to select, Ctrl to shoot, would be nice to have different control settings

**Comment by Wrzlprnft** [comment]
well, there is the key preference text file. we'd kust need to draw attention to it.

**Comment by LocalGod79** [comment]
https://groups.google.com/forum/#!category-topic/endless-sky/CEN6lAlmUn0

**Comment by Bladewood** [comment]
I've never read that thread before :|

**Comment by endless-sky** [comment]
If I got a couple of alternative key bindings that each had support from a lot of people (as opposed to a collection of many slightly different WASD bindings) it'd be pretty easy to just add buttons in the preferences to switch between them.

Note that most games that use WASD do so because you need your right hand for something else that's more important (yeah, I know, bias against left-handed people). In Endless Sky, on the other hand, fine control of your ship's movement is far more important than anything else, so it makes sense to make the motion controls right-handed by default. Plus, this is an Escape Velocity clone, and that's what they did:

![controls](https://cloud.githubusercontent.com/assets/11450262/24279398/0b07dc84-1020-11e7-971f-5a9b220f12dd.png)

**Comment by KiLEdEnNis** [comment]
a lot of people use their mouse for stuff on planets (moons & stations too), the M & I screen and because of the tactical click stuff even mid-flight, and I bet a lot (including me( hold their mouse even when not in use

especially that tactical click support makes the arrow key controls very hard to use; your right hand are on the arrows, controls on the keyboard and you can use your mouse to control ships,
so maybe WASD should be the default controls to ensure everyone can use the mouse like that without having to edit their key bindings
(and new players might get confused, thinking they're not supposed to use the mouse as well)

**Comment by gunqqer** [comment]
Oh the whole left handing thing I am left handed and the current control scheme works fine. I do mostly use it like a right handed person with WASD but arrow keys are good for left handed people 

(I'm left handed, AMA!)

**Comment by DingusShingleton** [comment]
I too am left handed, and the scheme to which I am referring could easily have a right handed version* as well (Arrows to move, WASD to target/fire). I personally eschew using the mouse while flying because I find its functionality limited; I only use it when I am on planet, since my right hand no longer needs to be on the arrows and my left can stay put to utilize some shortcuts.

However, even having said that, I find that when I *do* use the mouse to fly and my left hand is on the arrows.... all the hotkeys are halfway across the keyboard! So IMO, KB/Mouse setup makes more sense using WASD to navigate anyway, since WASD is nestled amongst far more usable hotkeys than are the arrows. I'd also add that basically every major FPS for the last, I don't know, 15 years? Has used WASD for control, and I have never heard of a right handed person having trouble with that, for what it's worth.

*But really, my setup could be the left-handed version of the standard righty WASD layout XD

**Comment by Jorlem** [comment]
I'd just like to state my support for the current setup, and my hope that WASD does not become the default, and if it is added, it is as an alternate setup at most.  

Personally, I cannot stand WASD control schemes, and when I try to use them half the time I end up hitting the wrong keys, something that isn't even slightly an issue with the arrow keys.  I just don't have the needed dexterity with my left hand (which is amusing in itself, as the word dexterity has its roots in referring to the right hand).

**Comment by DingusShingleton** [comment]
Nowhere did anyone suggest changing the default scheme. The idea was always to have alternate setups available in addition to the standard one.

**Comment by Jorlem** [comment]
KiLEdEnNis did bring up the idea that WASD should maybe be the default, which is what I was objecting to.  Sorry for not specifying.

**Comment by VonZorch** [comment]
This doesn't need a fix.  It's easy to do.  
Just copy the keys.txt file to something like Default Keys.bak then change key bindings.
copy the new keys.txt to My Keys.bak
This can be repeated as desired.
Then to switch delete keys.txt, copy the desired key file and rename to keys.txt
You need to restart the game but that's not a big deal.

**Comment by Wrzlprnft** [comment]
> You need to restart the game but that's not a big deal.

i am sorry to tell you, but it actually is. it is one of the major cons in steam reviews this game gets, the “unintuitive“ default controls. further, most of those people do not rebind them on their own, neither look up and find out about keys.txt and where to get alternatives from. the point “the default it comes with is crap, and the game does not offer any alternative itself. i should not have to rebind them or look online“.
given that our game does in no mean require online interaction with the community elsewhere, i support the idea of having different key-bindings as a default to choose from (like in halo, as is mentioned somewhere above).

**Comment by Tadrix** [comment]
Why exactly aren't we giving players the option to change key bindings themselves in the game, with the "Default" option restoring current defaults? It would be so hard to implement? If so, why?

**Comment by jafdy** [comment]
You can change the key bindings in settings.

**Comment by Tadrix** [comment]
Yep, you're right (I use default settings, so on seeing this assumed the only way to change those is keys.txt editing)

What's the problem then? Steam people don't know how to use this feature? Only thing it misses is "restore default bindings".

**Comment by Bladewood** [comment]
First impressions are generally the biggest.

**Comment by KiLEdEnNis** [comment]
when I started the game I immediately went to settings to change the keys, some Steam users aren't that... experienced, so to say
maybe when the game is first launched (or just doesn't have a specified key setup), the game could ask if you want arrows or WASD controls & tell you you can change it in the settings?
seems a bit much, but could work

**Comment by Tadrix** [comment]
```python
help "basics 0"
	`Use [Up] to accelerate, [Down] to decelerate, and [Left] and [Right] arrow keys to turn your ship. You can land by pressing [L]. These key bindings can always be changed in the Preferences menu.`
```

**Comment by Wrzlprnft** [comment]
yes. i don't think it's the arrow keys which are the issue, it's the rest, or the kombination.

On the other hand, some people don't read those. for real. and give a negative review afterwards.

different presets, or defaults to return to, would be just a quality of life thing for many, and i like the idea.

**Comment by Tadrix** [comment]
To be precise, game currently doesn't have such a tip. My "quote" was an improvisation. I'm suggesting to create one. It would at least partially help (players who read tips), while we are thinking of alternative schemes.

The bad news? If a player ignores the existence of tips and Preferences menu, (s)he will ignore the option to change entire schemes also (because they most likely will be in the same Preferences menu). Even throwing these options in their face could fail in the case of Next > Next > Ok mindset. Seriously, the need to come up with so much as a name for their character is an onerous inconvenience for some. The question is if we are aiming at such audience. They could also ignore the better part of lore and story, then complain that Endless Sky's setting has no depth.

**Comment by endless-sky** [comment]
Honestly speaking, anyone who can't figure out that you can click on "Preferences" to view or change key bindings, probably isn't going to be able to figure out even the basics of ship mechanics or trading or missions, either. So it's just screening them out a little bit earlier in the process. :)

**Comment by DingusShingleton** [comment]
I feel that there is a middle ground of people who might see the default controls as unintuitive, but find the prospect of remapping every key before they even know what most of them do to be a tad daunting. For those people, the ease of selecting a keyset by description ("Left handed, mouse", or "right handed, no mouse", just for example) would be a huge help. Personally, I would have made good use of this feature had I been able to, and it would have enhanced my experience by a certain amount.

On that note, I certainly do think this is a pretty minor QoL feature that should be on a distant backburner somewhere, if MZ even thinks it's worth the effort. If it makes it into 1.0.0, swell. If not, oh well.

**Comment by KiLEdEnNis** [comment]
to be fair MZ has a great point (& it's basically my opinion too); some Steam plebbies start the game, can't even figure out how the simple menu works, don't bother trying, leave a bad review and quit the game

although because of the new zooming & tactical clicking thing, the game "assumes" ya have a hand for the keyboard commands, 1 for the arrow keys & 1 for the mouse
I wouldn't change it for those dumb Steam users, I would change it so the default setup allows you to use everything: 1 hand for moving & commands, 1 hand for the mouse
